fix countermap to prevent data race

hannahkm opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Commit and PR titles should be prefixed with the general area of the pull request's change.

-->
### What does this PR do?

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

Fixes a data race introduced in #3023 that appears during concurrent `Inc()` and `GetAndReset()` calls.

### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
* If this resolves a GitHub issue, include "Fixes #XXXX" to link the issue and auto-close it on merge.
-->

In the old implementation, a race could occur if:

1. Inc() function calls Load(key)
2. The range function in GetAndReset() could reach and LoadAndDelete(key) since the range function itself is not atomic (key/values can change while Range is iterating)
3. GetAndReset() the current value associated with key
4. Inc() increments the value associated with key, but it has already been returned and deleted, so this count is lost forever

The race only occurs when multiple `Inc()` calls are happening concurrently with one or more `GetAndReset()` calls. While this is unlikely to happen in the wild, it _does_ cause data loss and flaky tests. 

### Reviewer's Checklist
<!--
* Authors can use this list as a reference to ensure that there are no problems
  during the review but the signing off is to be done by the reviewer(s).
-->

- [ ] Changed code has unit tests for its functionality at or near 100% coverage.
- [ ] [System-Tests](https://github.com/DataDog/system-tests/) covering this feature have been added and enabled with the va.b.c-dev version tag.
- [ ] There is a benchmark for any new code, or changes to existing code.
- [ ] If this interacts with the agent in a new way, a system test has been added.
- [ ] New code is free of linting errors. You can check this by running `golangci-lint run` locally.
- [ ] Add an appropriate team label so this PR gets put in the right place for the release notes.
- [ ] Non-trivial go.mod changes, e.g. adding new modules, are reviewed by @DataDog/dd-trace-go-guild.


Unsure? Have a question? Request a review!

kakkoyun commented on internal/utils.go:
```diff
@@ -72,36 +72,120 @@ func (l *LockMap) Get(k string) string {
+
+func (cm *CounterMap) GetAndReset() map[string]int64 {
+	// Ensure only one GetAndReset operation runs at a time.
+	cm.mu.Lock()
```
```suggestion
	if cm.mu.TryLock() {
		// Someone else is already reporting!
		return nil
	}
```

kakkoyun commented on internal/utils.go:
```diff
@@ -72,36 +72,120 @@ func (l *LockMap) Get(k string) string {
-		if ok {
-			ret[key] = v.Value()
+func (cm *CounterMap) Inc(key string) {
+	// Mark operation as in-flight.
+	cm.inFlightOps.Add(1)
+
+	// Get the current active state.
+	state := cm.activeState.Load()
+
```
```suggestion
    // THIS WON'T COMPILE!!
	state := cm.activeState.Load()
	state.cond.L.Lock()
	state.inFlightOps.Add(1)		
	state.cond.L.Unlock()
```

kakkoyun commented on internal/utils.go:
```diff
@@ -72,36 +72,120 @@ func (l *LockMap) Get(k string) string {
+	for cm.inFlightOps.Load() > 0 {
+		cm.cond.Wait() // Releases the lock while waiting.
+	}
+	cm.cond.L.Unlock()
```
```suggestion
	defer cm.cond.L.Unlock()
```
Now that I think, we shouldn't allow any more writes when we start reporting.

kakkoyun commented on internal/utils.go:
```diff
@@ -72,36 +72,120 @@ func (l *LockMap) Get(k string) string {
+		// Remove operation from in-flight list.
+		// If this was the last in-flight operation, signal waiters
+		if cm.inFlightOps.Add(-1) == 0 {
+			cm.cond.L.Lock()
+			cm.cond.Signal()
+			cm.cond.L.Unlock()
```
```suggestion
			// It is allowed but not required for the caller to hold c.L during the call.
			// See: https://pkg.go.dev/sync#Cond.Signal
			cm.cond.Signal()
```

kakkoyun reviewed (commented):
As we reviewed online and discovered some issues. I'm adding some notes.

Thanks for the tedious work on this.

kakkoyun commented on internal/utils.go:
```diff
@@ -72,36 +72,125 @@ func (l *LockMap) Get(k string) string {
+	defer func() {
+		// Remove operation from in-flight list.
+		// If this was the last in-flight operation, signal waiters
+		if state.inFlightOps.Add(-1) == 0 {
+			// It is allowed, but not required for the caller to hold c.L during the call.
+			// See: https://pkg.go.dev/sync#Cond.Signal
+			state.cond.Signal()
+		}
```
```suggestion
		state.cond.L.Lock()
		if state.inFlightOps.Add(-1) == 0 {
			// It is allowed, but not required for the caller to hold c.L during the call.
			// See: https://pkg.go.dev/sync#Cond.Signal
			state.cond.Signal()
		}
		state.cond.L.Lock()		
```

kakkoyun commented on internal/utils.go:
```diff
@@ -72,36 +72,125 @@ func (l *LockMap) Get(k string) string {
+	for oldState.inFlightOps.Load() > 0 {
+		oldState.cond.Wait() // Releases the lock while waiting.
+	}
+	oldState.cond.L.Unlock()
```
```suggestion
	defer oldState.cond.L.Unlock()
```

felixge commented:
> The race only occurs when multiple Inc() calls are happening concurrently with one or more GetAndReset() calls. While this is unlikely to happen in the wild, it does cause data loss and flaky tests.

I'm probably missing some context. Could you please describe a specific goroutine execution scenario where the race occurs?

hannahkm commented:
@felixge In the old implementation, we could have an ordering like:

1. Inc() function calls `Load(key)`
2. The range function in GetAndReset() could reach and `LoadAndDelete(key)` since the range function itself is not atomic (key/values can change while Range is iterating)
3.  GetAndReset() the current value associated with key
4. Inc() increments the value associated with key, but it has already been returned and deleted, so this count is lost forever

It doesn't happen often, but it has been causing our tests to flake sometimes. I've added a new test (`TestCounterMap/concurrent_with_reset`) that should catch the race more frequently, especially if run with a `-count=100` or something similar.

darccio commented:
@hannahkm Do you remember why this PR is still in draft?
